### PR TITLE
Add some valued variable names

### PR DESCRIPTION
**Pull request by vvscode**
I just rename params inside insert-templates

And start to extract helper functions into separate file to improve testing.

@ewnd9 

- What about typescript/flow/etc. ?
- Is there any `describe` alternative in ava? ( I didn't find )
- What about jsdoc ( I ask about common format for documentation  -- I mean places like here https://github.com/ewnd9/anki-apkg-export/pull/7/files#diff-1fdf421c05c1140f6d71444ea2b27638R35 -- I preffer give as much info as I can, but it's better to have jsdoc template )
- What about asserts in code to check input data? 

We can discuss that questions not here (issue / some other place ). But I'm interested in

### CODE REVIEW

File: test/helpers.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+  getMedia,
+  getSave,
+  getSql,
+  getTemplate,
+  getZip,
+  rand
+} from '../src/helpers';
+
+test('checksum', t => {
+  t.plan(2);
```
Comment (ewnd9) [inline test/helpers.js]: Please read [when-to-use-plan.md](https://github.com/avajs/ava/blob/master/docs/recipes/when-to-use-plan.md)

**Comment by ewnd9** [comment]
I really appreciate your effort, but, sorry, it feels wrong to me, especially 

```js
  /**
   * @param front
   * @param back
   */
  const addCard = getAddCard(
    update,
    top_deck_id,
    top_model_id,
    SEPARATOR
  );
```

Why don't just use `es6` classes or old prototypes-based?

> Is there any describe alternative in `ava`? ( I didn't find )

Why? What is the advantages? 

> What about `typescript`/`flow`/etc. ?

I like `flow`, haven't used it much though. We can add support for both since they support it via external files

> What about `jsdoc`

I don't like verbosity of `jsdoc`. I mean, of course, you need at least something as a comment in your function above, but I hope you will see redundancy in `class`-es version.

> What about asserts in code to check input data?

I use [`tcomb`](https://github.com/gcanti/tcomb) for, what I call, critical type check (validation before saving to nosql db or processing server requests). In other cases it's better to implement `flow` or `typescript`.

**Comment by ewnd9** [comment]
Also, what really would be nice, it's to separate `webpack`-agnostic logic and use the [`browser`](https://gist.github.com/defunctzombie/4339901) property. Something like:

```js
// node.js
const sql = require('sql.js');
const fs = require('fs');
const template = fs.readFileSync(__dirname + '/../template.sql', 'utf-8');

export default function(deckName) {
  return new ApkgExporter({ deckName, sql, template });
}
``` 

```js
// browser.js
require('script!sql.js');
const sql = window.SQL;
const template = require('!raw!./../template.sql');

export default function(deckName) {
  return new ApkgExporter({ deckName, sql, template });
}
```

**Comment by vvscode** [comment]
> Why don't just use es6 classes or old prototypes-based?
Looks like misunderstanding. The goal for such construction - split code and add ability to pass dependencies from outside. `class` keep it state. functions - doesn't. 
```javascript
 const addCard = getAddCard(
    update,
    top_deck_id,
    top_model_id,
    SEPARATOR
  );
```
is just example of usage high order function. And I don't see nothing common with es6 classes / prototypes. I don't create class. I just use curing to increase testability

> Why? What is the advantages?
Right now I have `helpers.js` with number of functions. I don't like have large tests with more than 1-3 assertions. So `describe` for me - just way to group list of tests. Or I can describe tests inside test function? Just to have small tests in groups instead of current large tests

P.S. I'll be glad to discuss details in Russian, cause I'm not sure if my english cause infelicity

File: src/helpers.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+export const getMedia = () => {
+  const content = [];
+  return {
+    getContent: () => content,
+    addMedia: (filename, data) => content.push({filename, data})
+  };
+};
+
+export const getCssTemplate = () => {
```
Comment (ewnd9) [inline src/helpers.js]: Why it should be in the separate file? 

```js
`.card {		
  font-family: arial;		
  font-size: 20px;		
  text-align: center;		
  color: black;		
}`
```

File: src/helpers.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    return prev;
+  }, {});
+
+  zip.file('collection.anki2', new Buffer(binaryArray));
+  zip.file('media', JSON.stringify(mediaObj));
+
+  media.forEach((item, i) => zip.file(i, item.data));
+
+  if (process.env.APP_ENV === 'browser') {
+    return zip.generateAsync(Object.assign({}, { type: 'blob' }, options));
```
Comment (ewnd9) [inline src/helpers.js]: Is it possible to solve with `typeof`?

**Comment by ewnd9** [comment]
Let's try from another angle, what you don't like in the code below in terms of modularity and testability?

```js
function ApkgExporter({ deckName, ...options }) {
  this.options = {
    name: deckName,
    model_name: deckName,
    css: ` .card {
      font-family: arial;
      font-size: 20px;
      text-align: center;
      color: black;
    }`,
    ...options
  };

  this.db = new sql.Database();
  this.db.run(options.template);

  // ...
}


ApkgExporter.prototype.addCard = function(front, back) {

};

ApkgExporter.prototype.addMedia = function(filename, data) {

};
```

I see the logic in your way but it is just hard to read such code in compassion with classes 

> P.S. I'll be glad to discuss details in Russian, cause I'm not sure if my english cause infelicity

I don't see problems, let's please continue in English.

> just way to group list of tests

I'm afraid the only way to group tests in `ava` is to spread them in separate files

---

I edited your comment to add new line characters after quotes

```md
> Quote

Answer
```

**Comment by vvscode** [comment]
I see. 
https://github.com/vvscode/js--anki-apkg-export/pull/2
is what you like. 

Ok, I'll migrate changes

**Comment by ewnd9** [comment]
```js
  const flags = {};
  const dbMock = { export: () => flags['db.export'] = 'Some data' };
  const zipMock = {
    file: (path, content) => flags[`zip.file(${path})`] = content,
    generateAsync: options => flags[`zip.generateAsync`] = options
  };
  const mediaMock = [];
  const save = getSave(zipMock, dbMock, mediaMock);

  // it should process media by reference
  mediaMock.push({filename: '1.jpg'}, {filename: '2.bmp'});

  save({some: 'options', should: { be: 'here'}});
  const flagsKeys = Object.keys(flags);
  t.truthy(flagsKeys.includes('db.export'), 'should call .export on db');
  t.truthy(flagsKeys.includes('zip.file(collection.anki2)'), 'should save notes/cards db');
  t.truthy(flagsKeys.includes('zip.file(media)'), 'should save media');
  t.truthy(flagsKeys.includes('zip.file(0)'), 'should save media with two files');
  t.truthy(flagsKeys.includes('zip.file(1)'), 'should save media with two files');
  t.truthy(flagsKeys.includes('zip.generateAsync'), 'should call zip.generateAsync');
```

Such way of testing is more clear with a combination of [`proxyquire`](https://github.com/thlorenz/proxyquire) (mocking requires) and [`sinon`](https://github.com/sinonjs/sinon)  (as a proxy to inspect passed arguments in each call)

File: src/helpers.js (hunk 3 above)
Comment (vvscode) [inline src/helpers.js]: I miss something? or you ask about browser detection?

File: src/helpers.js (hunk 2 above)
Comment (vvscode) [inline src/helpers.js]: I suppose we need provide way to override / extend styles. And file -- we can have syntax hightlighting etc. 

doesn't look good for you?

File: src/helpers.js (hunk 3 above)
Comment (ewnd9) [inline src/helpers.js]: I see now it is a copy of my code :grin: I am thinking about ridding of `process.env.APP_ENV === 'browser'`. Probably, should be also passed in the separate file (https://github.com/ewnd9/anki-apkg-export/pull/7#issuecomment-263780460)

File: src/helpers.js (hunk 2 above)
Comment (ewnd9) [inline src/helpers.js]: It's 5 lines, too small to justify adding complexity. Overriding would be better as a constructor argument (https://github.com/ewnd9/anki-apkg-export/pull/7#issuecomment-263785469) so users will decide by themselves how they want to use their css

**Comment by vvscode** [comment]
@ewnd9  not sure about the idea use `proxyquire`

also I'll open new PR to make easier to read discussion 

https://github.com/ewnd9/anki-apkg-export/pull/8

**Pull request closed without merging**